A58-xds-stateful-session-affinity-weighted-clusters.md

```diff
@@ -0,0 +1,270 @@
+cluster, there is no guarantee the new cluster still exists at that point.
+So we need to adjust the refs in the cluster map for proper cluster lifecycle
+management.
+
+#### Changes to Config Selector
+
+ * Make the cluster map in config selector ref-counted. It enables
+   maintaining a ref to a swapped out cluster map if an xDS update happens
+   while processing an RPC.
+
+ * config selector takes a ref to the cluster map and passes it via a new
+   call-attribute (let’s say `XdsClusterMapAttribute`) to downstream entities.
+   The selected cluster-name is passed - as before - but without taking a ref
+   to the cluster entry.
+
+ * config selector also exposes the currently selected route entry via a new
+   call-attribute (let’s say `XdsRouteAttribute`).
+
+   * We also make the entire route-config ref-counted by adding it to the same
+     object as the cluster map.
+
+With the above changes the config selector makes 4 values available to the
+`StatefulSessionFilter` (and other filters): selected cluster, current route,
+cluster-map and the entire route-config.
+
+### `StatefulSessionFilter` Changes
+
+#### Response Side Processing
+
+In the filter's response processing, we set the cookie in the response, if the
+corresponding request didn’t have a cookie or the actual cluster+backend that
+processed the request didn’t match what the cookie had. The logic in pseudo-code
+is:
+
+```
+if (!upstreamAddress.has_value() || hostAddress != upstreamAddress) {
+  String clusterUsed = cluster name from XdsClusterAttributeTypeName
+  String toEncode = hostAddress + ";\"" + clusterUsed + "\"";
+  String encoded_address = base64_encode(toEncode);
+  Create a Cookie from our cookie configuration;
+  Set cookie's value as encoded_address;
+  Add the cookie header to the response;
+}
+```
+
+#### Request Side Processing
+
+The filter needs to process the cluster name present in the cookie.
+
+ * Extract cluster name from RPC’s cookie and validate it:
+
+   * Verify the cluster name is present in the current route (remember it is
+     exposed via `XdsRouteAttribute`). Route-action can only be single cluster
+     or weighted-clusters and cluster should match/be present in one of those.
+
+ * If validation succeeds, then override the cluster call-attribute (originally
+   set by the config selector) with the cluster name.
+
+ * Set the override-host value with the IP:port value from the cookie (this is
+   the current logic).
+
+### New `ClusterSelectionFilter`
+
+Because of the changes we made to the config selector we need a new filter
+to finalize cluster selection and to release refs to the cluster map and
+route config. This new filter is called `ClusterSelectionFilter` and it runs
+just before the Router filter. This is an internal implementation detail of
+gRPC so the xDS config is not aware of this new filter. Note:
+
+ * This is the second-last filter (retry/router filter being last).
```

Comment:
I am not too familiar with the internals of this logic so pardon for potentially dumb question here. Why does it need to run just before the router filter? Wouldn't we expect the HTTPRoute logic to need to pick the cluster according to route configuration first and then only if we have an overwrite value from the session cookie (as part of StatefulSessionFilter) it would need to be modified by ClusterSelectionFilter.
Perhaps the notion here is to let StatefulSessionFilter provide the cluster value from the cookie (if such is available) and store it in the call attributes and only if not set then Route would pick the cluster according to weights configuration?

Reply:
In the xDS data model, the workflow is as follows:
1. We choose the route and cluster based on the RouteConfig.
2. We iterate over the HTTP filters, executing each one in order.  Each filter can access the currently chosen route and cluster, so it can do things like get per-route config overrides, get which cluster the request will be sent to, or (in the case of the StatefulSession filter) override the cluster that the request will be sent to.
3. The last HTTP filter in the chain is the router filter, which is responsible for supporting retries and for sending the request to the chosen cluster.

In gRPC, unlike Envoy, the router filter doesn't actually exist as a distinct piece of code, because gRPC already has retry functionality built into the client channel, and we handle clusters in the LB policy tree.  So when we see the router filter in the xDS config, we just treat it as shorthand to tell us to use that existing functionality.

The new ClusterSelectionFilter being added here is basically just another step that will occur right before the retry code, so it will occur logically as the very first part of step 3.  Note that this filter is an internal implementation detail in gRPC, not something that will be exposed in the xDS API, and it will always be executed right before the retry code, so you can think of it as part of the router filter implemention in gRPC.  And despite its name, this filter is not actually selecting which cluster to send the request to; that decision is still being made in step 1, and the only thing this filter is doing is handling some ref-counting details that we need internally.

I hope that answers your question.  Please let me know if I haven't explained that in a clear way.

Reply:
Thank you @sanjaypujare and @markdroth